### PR TITLE
os/bluestore: fix AU accounting in bluestore_cache_other mempool.

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/56424

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/os/bluestore/bluestore_types.cc [hunk 1]
```diff
@@ -385,29 +386,40 @@ bluestore_blob_use_tracker_t::operator=(const bluestore_blob_use_tracker_t& rhs)
...
-  bytes_per_au = new uint32_t[num_au];
+  ceph_assert(au_count != 0);
+  ceph_assert(num_au == 0);
+  ceph_assert(alloc_au == 0);
+  num_au = alloc_au = au_count;
+  bytes_per_au = new uint32_t[alloc_au];
   mempool::get_pool(
     mempool::pool_index_t(mempool::mempool_bluestore_cache_other)).
-      adjust_count(1, sizeof(uint32_t) * num_au);
+      adjust_count(alloc_au, sizeof(uint32_t) * alloc_au);
```
Comment (aclamk) [inline src/os/bluestore/bluestore_types.cc:407]: I think we here add 1 object size 4*alloc_au to mempool.

File: src/os/bluestore/bluestore_types.cc [hunk 2]
```diff
@@ -385,29 +386,40 @@ bluestore_blob_use_tracker_t::operator=(const bluestore_blob_use_tracker_t& rhs)
...
 
   for (uint32_t i = 0; i < num_au; ++i) {
     bytes_per_au[i] = 0;
   }
 }
 
+void bluestore_blob_use_tracker_t::release(uint32_t au_count, uint32_t* ptr) {
```
Comment (aclamk) [inline src/os/bluestore/bluestore_types.cc:414]: +1 for adding dedicated release function

File: src/os/bluestore/bluestore_types.h [hunk 3]
```diff
@@ -247,18 +247,19 @@ struct bluestore_blob_use_tracker_t {
   //   1) Struct isn't packed hence it's padded. And even if it's packed see 2)
   //   2) Mem manager has its own granularity, most probably >= 8 bytes
   //
-  uint32_t au_size; // Allocation (=tracking) unit size,
-                    // == 0 if uninitialized
-  uint32_t num_au;  // Amount of allocation units tracked
-                    // == 0 if single unit or the whole blob is tracked
+  uint32_t au_size;  // Allocation (=tracking) unit size,
+                     // == 0 if uninitialized
+  uint32_t num_au;   // Amount of allocation units tracked
```
Comment (aclamk) [inline src/os/bluestore/bluestore_types.h:252]: Do we want to save some mem space by change num_au and alloc_au to uint16_t ?
Is 24->16 bytes worth it ?

**Review by aclamk — APPROVED** [review]
Good fix on inconsistent mempool tracking.

File: src/os/bluestore/bluestore_types.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/bluestore_types.cc:407]: just modified this a bit so each entry is treated as an object. The original plan was to be able to shrink this array if needed and hence adjust object count as well. But finally I've redesigned this stuff a bit differently.

**Comment by NitzanMordhai** [comment]
https://pulpito.ceph.com/yuriw-2022-07-09_14:48:58-rados-wip-yuri6-testing-2022-07-07-1429-distro-default-smithi/

Failures, unrelated:
6921591 - https://tracker.ceph.com/issues/52321
6921593 - https://tracker.ceph.com/issues/52321
6921594 - https://tracker.ceph.com/issues/55789
6921595 - https://tracker.ceph.com/issues/52321
6921596 - https://tracker.ceph.com/issues/52321
6921597 - https://tracker.ceph.com/issues/55789

Details:
6921591\6921593\6921595\6921596 - qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Rook
6921594\6921597 -  cls_rgw.index_suggest_complete assert failuer during test

**Pull request merged**